### ISSUE TITLE
[PydanticV2] Add parameter to use Python Regex Engine in order to support look-around

### ISSUE THREAD

**Issue by ilovelinux** [koxudaxi/datamodel-code-generator#2232]
**Is your feature request related to a problem? Please describe.**
1. I wrote a valid JSON Schema with many properties
2. Some properties use patterns with look-ahead and look-behind inside, which are supported by JSON Schema specifications.
    See: [JSON Schema supported patterns](https://json-schema.org/understanding-json-schema/reference/regular_expressions)
4. `datamodel-code-generator` generated the PydanticV2 models.
5. PydanticV2 doesn't support look-around, look-ahead and look-behind by default (see https://github.com/pydantic/pydantic/issues/7058) 
```Python traceback
ImportError while loading [...]: in <module>
[...]
.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py:205: in __new__
    complete_model_class(
.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py:552: in complete_model_class
    cls.__pydantic_validator__ = create_schema_validator(
.venv/lib/python3.12/site-packages/pydantic/plugin/_schema_validator.py:50: in create_schema_validator
    return SchemaValidator(schema, config)
E   pydantic_core._pydantic_core.SchemaError: Error building "model" validator:
E     SchemaError: Error building "model-fields" validator:
E     SchemaError: Field "version":
E     SchemaError: Error building "str" validator:
E     SchemaError: regex parse error:
E       ^((?!0[0-9])[0-9]+(\.(?!$)|)){2,4}$
E         ^^^
E   error: look-around, including look-ahead and look-behind, is not supported
```

**Describe the solution you'd like**
PydanticV2 supports look-around, look-ahead and look-behind using Python as regex engine: https://github.com/pydantic/pydantic/issues/7058#issuecomment-2156772918

I'd like to have a configuration parameter to use `python-re` as regex engine for Pydantic V2.

**Describe alternatives you've considered**
Workaround:

1. Create a custom BaseModel
```python
from pydantic import BaseModel, ConfigDict


class _BaseModel(BaseModel):
    model_config = ConfigDict(regex_engine='python-re')
```
2. Use that class as BaseModel:
```sh
datamodel-codegen --base-model "module.with.basemodel._BaseModel"
```

**Comment by dpeachey** [comment]
datamodel-code-genertor should automatically add `ConfigDict(regex_engine='python-re')` for you when it sees a field called `pattern` that is using  look-around, look-ahead or look-behind regex.

Are you using `collapse_root_models=True`? Because I am using this and I see that `ConfigDict(regex_engine='python-re')` is not added to `BaseModel` that have merged fields from a `RootModel` where the `RootModel` had ``ConfigDict(regex_engine='python-re')` `. So I am curious if this also the same use case for you, and if so then it may be a bug,

**Comment by ilovelinux** [comment]
@dpeachey I am!

I edited the issue and I added the full config file I'm using 🙂

**Comment by dpeachey** [comment]
Yeah then I think it is a bug. Ideally I think when using `collapse_root_models=True` the `ConfigDict` from the `RootModel` should be merged with the `ConfigDict` in the parent `BaseModel`.

**Comment by gaborbernat** [comment]
PR welcome.
